Charts: Refine helm installation

animeshk08 opened the pull request:
This commit updates helm installation
by using helm repos.

Signed-off-by: Animesh Kumar <animuz111@gmail.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind test
> /kind failing-test

> /kind flake

 /kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
<!-- 
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #283 

**Special notes for your reviewer**:


**Release note**:
```
Charts: Refine helm installation
```

animeshk08 commented on charts/index.yaml:
```diff
@@ -0,0 +1,76 @@
+    created: "2020-05-25T02:02:09.483531499+05:30"
+    description: Azure File Container Storage Interface (CSI) Storage Plugin
+    digest: ad581478836e360fc6b8367791867e7184268857b4f50b322ef5245010768bd4
+    home: https://github.com/kubernetes-sigs/azurefile-csi-driver
```
Hi @andyzhangx, I have not included maintainer details for now. 
```
maintainers:
    - email: <email>
      name: <username/name>
```

Let me know if you want me to include your details. Thanks :)

andyzhangx replied:
here is the maintainer details, thanks.
```
maintainers:
    - email: xiazhang@microsoft.com
      name: andyzhangx
```

andyzhangx commented:
/ok-to-test

andyzhangx commented:
BTW, how can I install old release, e.g. in this PR, how can I install `v0.7.0` release, pls provide instruction in this PR, thanks.

animeshk08 commented:
Hi, @andyzhangx Let me know if the project structure looks good to you now. I have updated the maintainer details and also added in the documentation info about how to install a specific version of the charts.

I have a few comments/suggestion:
* The version of some of the charts(`0.2.0` and `0.3.0`) is not prepended by 'v', i.e. `v0.2.0` and `v0.3.0`. This results in inconsistency in the name of the package created. Let me know if you want me to update that.

* By default if the version is not specified the latest version present in `index.yaml` is installed. In this case, this will be `v0.7.0`. I understand this is not the development/latest version. Now, we can move in two paths to include the latest version.
1. Add separate instruction of installing development version and not include it in `index.yaml` e.g. https://github.com/elastic/helm-charts/tree/master/elasticsearch#installing

2. Create a package of the latest version as well and include it in `index.yaml`. This is followed in most Azure repos AFAIK, e.g. https://github.com/Azure/aad-pod-identity/tree/master/charts. If we follow this approach I suggest changing the version of the development charts from `latest` to `v0.8.0` or any other preferred version.

Let me know what are your views. I will update the PR accordingly. Thanks :)

andyzhangx commented:
@animeshk08 we should only supports latest 3 (or less) versions in this PR, started from `v0.4.0`.
I think we could add a new parameter `--version` in helm, and if not specified, install `latest` version by default. So on master branch, it should always use `latest` version.

e.g. for aad-pod-identity:
```console
helm install aad-pod-identity/aad-pod-identity --name pod-identity --version 1.5.4
```
https://github.com/Azure/aad-pod-identity/blob/master/charts/index.yaml#L18

animeshk08 commented:
> I think we could add a new parameter --version in helm, and if not specified, install latest version by default. So on master branch, it should always use latest version.

I am sorry if I was not clear before. Such a parameter `--version` is already in helm and I have also mentioned its use-case in the docs, What I wanted to tell you was that if you look at the repo for r aad-pod-identity. (https://github.com/Azure/aad-pod-identity/tree/master/charts). Their latest charts version is 0.2.0.
https://github.com/Azure/aad-pod-identity/blob/master/charts/aad-pod-identity/Chart.yaml#L4

Now, they have created a package for the latest version as well.(https://github.com/Azure/aad-pod-identity/blob/master/charts/aad-pod-identity-2.0.0.tgz). and referenced the same in their `index.yaml`.(https://github.com/Azure/aad-pod-identity/blob/master/charts/index.yaml#L18).

Now if we follow the same, and I create a package for the latest version and include it the `index.yaml` for our charts, my only concern is that our package will look like this: `azurefile-csi-driver-latest.tgz`. and if I continue using the version `latest` in `index.yaml` as well. Running  commands like `helm search -l azurefile-csi-driver` we will get the packages as:
```
NAME                                            CHART VERSION   APP VERSION     DESCRIPTION 
azurefile-csi-driver/azurefile-csi-driver       latest          latest          Azure File Container Storage Interface (CSI) Storage Plugin
azurefile-csi-driver/azurefile-csi-driver       v0.6.0          v0.6.0          Azure File Container Storage Interface (CSI) Storage Plugin
azurefile-csi-driver/azurefile-csi-driver       v0.5.0          v0.5.0          Azure File Container Storage Interface (CSI) Storage Plugin
azurefile-csi-driver/azurefile-csi-driver       v0.4.0          v0.4.0          Azure File Container Storage Interface (CSI) Storage Plugin

```
As you can see the version is `latest` which does not tell much and also stands out. My suggestion was to drop using `latest` as the version instead provide it with a version number which in our case will be `v0.8.0`. This suggestion comes after seeing other helm repos. As you can see they have a specified chart version and not `latest`. Our charts are also going to be public and I think we should follow the same
```
NAME                    CHART VERSION   APP VERSION     DESCRIPTION                                                 
stable/jasperreports    7.0.11          7.2.0           DEPRECATED The JasperReports server can be used as a stan...
stable/artifactory      7.3.1           6.1.0           DEPRECATED Universal Repository Manager supporting all ma...
stable/artifactory-ha   0.4.1           6.2.0           DEPRECATED Universal Repository Manager supporting all ma...
stable/chartmuseum      2.12.2          0.12.0          Host your own Helm Chart Repository                         
stable/dmarc2logstash   1.3.0           1.0.3           Provides a POP3-polled DMARC XML report injector into Ela...
stable/pgadmin          1.2.3           4.18.0          DEPRECATED - moved to new repo, see source for new location 
stable/satisfy          1.0.0           3.0.4           Composer repo hosting with Satisfy                          
stable/sentry           4.2.1           9.1.2           Sentry is a cross-platform crash reporting and aggregatio...
stable/sonatype-nexus   1.23.1          3.20.1-01       DEPRECATED - Sonatype Nexus is an open source repository ...
azurefile-csi-driver/
azurefile-csi-driver    latest          latest          Azure File Container Storage Interface (CSI) Storage Plugin
```


If you don't want this we can stick with `latest` as the version only.

> we should only supports latest 3 (or less) versions in this PR, started from v0.4.0

I will update the PR to start from `v0.4.0` :)

animeshk08 commented on charts/README.md:
````diff
@@ -6,12 +6,22 @@ Quick start instructions for the setup and configuration of azurefile CSI driver
+### Install a specific version of chart
+Specify the version of the chart to be installed using the `--version` tag. 
+```console
+helm install --name azurefile-csi-driver azurefile-csi-driver/azurefile-csi-driver --namespace kube-system --version v0.6.0
````
I have included the use of `--version` here.

andyzhangx commented:
thanks, let's use `v0.8.0` instead of `latest`, also could you fix this spelling error:
```
Checking spelling...
error: deploy/example/snapshot/README.md:5:3: "Minumum" is a misspelling of "Minimum"
Found spelling errors!
```

animeshk08 commented:
Yes, I am on it.

animeshk08 commented:
@andyzhangx PR is ready for review.

andyzhangx commented on charts/latest/azurefile-csi-driver/Chart.yaml:
```diff
@@ -1,5 +1,5 @@
 apiVersion: v1
-appVersion: latest
```
after second thought, I think we should keep `latest` tag, `latest` means current master branch version, and if we specify `v0.8.0`, we need to change this version num every time there is new release.

andyzhangx replied:
sorry for the turn around, while I think using `latest` should be more appropriate for maintenance in the future.

animeshk08 replied:
Hi, @andyzhangx so we have a few other problems while using `latest` as the version.
Helm charts specify that the chart version has to be of SemVer2 format.
https://helm.sh/docs/topics/charts/#charts-and-versioning

This is needed to parse the version while using `--version` tag with `helm install`. Now the keyword `latest` does not follow this rule. Hence I am getting the following error:
```
Error: chart "azurefile-csi-driver" matching version "latest" not found in azurefile-csi-driver index. (try 'helm repo update'). improper constraint: latest
```
 When I try running
 ```
helm install --name azurefile-csi-driver azurefile-csi-driver/azurefile-csi-driver --namespace kube-system --version latest --debug
```
I confirmed this problem by checking these issues:
* https://github.com/helm/helm/issues/4109
* https://github.com/helm/helm/issues/3600

AFAIK there is also no way to specify a chart as default in `index.yaml`. What helm does is sort the chart versions and install the latest one if the `version` is not explicitly mentioned. 
https://github.com/helm/helm/blob/15d9a6190f5e3ed01b5f2f670be48e7a0290518d/pkg/repo/index.go#L152

Let me know how you would like me to move forward from here.

andyzhangx replied:
hi, what about remove `latest` in helm installation, so the biggest version in helm is `v0.7.0` , and keep charts\latest unchanged
CI test would run on chart `latest` , and don't change deploy version, keep it as `latest` ?

animeshk08 replied:
Hi, so you don't want me to include `latest` inside `index.yaml`? In this case, the default installation will be `v0.7.0`.  If that's the case. I would again suggest following what elastic does. https://github.com/elastic/helm-charts/tree/master/elasticsearch#installing. Include stable version installation via helm repo and development(latest) version installation directly.

andyzhangx commented on deploy/csi-azurefile-controller.yaml:
```diff
@@ -121,7 +121,7 @@ spec:
               cpu: 10m
               memory: 20Mi
         - name: azurefile
-          image: mcr.microsoft.com/k8s/csi/azurefile-csi:latest
```
keep `latest`

andyzhangx replied:
undo this change

andyzhangx commented on charts/README.md:
```diff
@@ -6,12 +6,22 @@ Quick start instructions for the setup and configuration of azurefile CSI driver
 1. [install Helm Client 3.0+ ](https://helm.sh/docs/intro/quickstart/#install-helm)
 
-## Install AzureFile via `helm install`
+## Install a released version of Azurefile using Helm repository
```
`Install CSI Driver master version using Helm repository`

andyzhangx commented on charts/latest/azurefile-csi-driver/Chart.yaml:
```diff
@@ -1,5 +1,5 @@
+appVersion: v0.8.0
 description: Azure File Container Storage Interface (CSI) Storage Plugin
 name: azurefile-csi-driver
-version: latest
```
undo this change

andyzhangx commented on charts/latest/azurefile-csi-driver/values.yaml:
```diff
@@ -1,7 +1,7 @@
 image:
   azurefile:
     repository: mcr.microsoft.com/k8s/csi/azurefile-csi
-    tag: latest
```
undo this change

andyzhangx commented on deploy/csi-azurefile-node-windows.yaml:
```diff
@@ -66,7 +66,7 @@ spec:
               cpu: 10m
               memory: 20Mi
         - name: azurefile
-          image: mcr.microsoft.com/k8s/csi/azurefile-csi:latest
```
undo this change

andyzhangx commented on deploy/csi-azurefile-node.yaml:
```diff
@@ -66,7 +66,7 @@ spec:
               cpu: 10m
               memory: 20Mi
         - name: azurefile
-          image: mcr.microsoft.com/k8s/csi/azurefile-csi:latest
```
undo this change

animeshk08 commented:
@andyzhangx PR is ready for review :)
